### PR TITLE
docs: add design for blocking CephCluster deletion

### PR DESCRIPTION
**Pull request by BlaineEXE**
Add a design for blocking CephCluster deletion when user data is
detected in the cluster to serve as additional protection against
accidentally deleting user data.

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+## Optimization
+A possible optimization we can make is to still continue deleting the Ceph cluster if pools do not
+contain user data. If there is no user data stored in the cluster, no data will be destroyed when
+the cluster is deleted.
+
+It would be most ideal to have an option in Ceph to disallow clients from writing new data once we
+get a delete request for a CephCluster. If we can set such an option, we can eliminate a race
+condition where we detect that a pool has no user data, then a user writes data into the pool, and
+then Rook proceeds to delete the CephCephcluster.
```
Comment (leseb) [inline design/ceph/block-ceph-delete-if-pool-data.md:28]:
```suggestion
then Rook proceeds to delete the Cephcluster custom resources.
```

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  - For each zone in `zones` (from JSON output):
+    - `radosgw-admin --rgw-zonegroup=$zonegroup --rgw-zone$zone zone get` to get an object store's buckets
+    - `radosgw-admin --rgw-zonegroup=$zonegroup --rgw-zone=$zone bucket list` to get buckets
+    - Ignore a single instance of a bucket named `rook-ceph-bucket-checker-*`
+    - If other buckets exist, assume the store is in-use
+
+##### Corner cases
+1. If a user creates an object store without using CephObjectStore and inside that store creates a
+bucket with a name in the format `rook-ceph-bucket-checker-*`, Rook will not stop cluster deletion.
+2. `.rgw.root` is present in Ceph pool list but not included in info from `radosgw-admin`... is this
```
Comment (leseb) [inline design/ceph/block-ceph-delete-if-pool-data.md:76]: Currently, it seems to be the case yes.

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+   item ALWAYS the same name?????? (TODO)
+
+##### Background technical information
+Zone name and zonegroup name are usually the same as CephObjectStore name except with new
```
Comment (leseb) [inline design/ceph/block-ceph-delete-if-pool-data.md:80]: I think there is a catch, if this is a multi-site cluster and the admin wants to remove one site (since data are synced to another zone), then it's probably acceptable to go with the deletion.

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+written any data.
+
+### Pool information in CR statuses
+For Rook resources that create pools, it would be good to list the pools on the CR status to help
+admins cross-referencing things during deletion (but also during normal Day 1 and Day 2 ops). List
+data and metadata pools separately. E.g., look at a CephFilesystem CR's status.pools.data to find a
+data pool for CephNFS to use (day 1).
+
+### CephClusters in external mode
+TODO: should we act differently in this case since the cluster exists merely to connect to an
```
Comment (leseb) [inline design/ceph/block-ceph-delete-if-pool-data.md:123]: The deletion will:

* affect CSI, since the cm might go away (TBV)
* affect OB/OBC since the CephObjectStore relies on the existence of CephCluster in some place IIRC

**Review by leseb — COMMENTED** [review]
We need to address the case where deletion is intentional and the cluster has data. Simply explaining that the finalizers need to be removed manually.

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
+# Blocking CephCluster deletion if user data is present
+
+
+## Use case
+As an administrator, I want to be protected from accidentally deleting user data if the CephCluster
+is deleted (accidentally or on purpose).
+
+
+## Determining if a Ceph is in use
```
Comment (travisn) [inline design/ceph/block-ceph-delete-if-pool-data.md:9]: After reading through this and pondering all our discussions in the last week, this is where I'm landing...
1. Each controller that owns creating/deleting Ceph resources would block removal if those resources appear to contain data
2. Each controller that has dependent controllers would block removal if any dependent CRs exist.

For example, this would translate into:
- CephCluster: block if any dependent CRs exist in its own namespace: CephBlockPool, CephObjectStore, CephFilesystem, CephNFS, ...
- CephObjectStore: block if any OBs exist (already implemented), or if any CephNFS CRs are referencing the object store
- CephBlockPool: block if any rbd images exist in the pool (already implemented)
- CephFilesystem: block if the filesystem appears in use, or if any CephNFS CRs are referencing the filesystem 

Of note is that the CephCluster doesn't directly check for any Ceph resources. By checking for dependent CRs which in turn own Ceph resources, we effectively block since there might be Ceph resources in use. If the user deletes those dependent CRs, then the CephCluster would be able to proceed with the removal.

I believe this keeps the implementation simple and also accomplishes the goals of 1) blocking uninstall when there is data and 2) not overly complicating dev scenarios.

Comment (BlaineEXE) [reply design/ceph/block-ceph-delete-if-pool-data.md:9]: Would you also  include not deleting RBD mirroring CRs if block pools reference them?
Or deleting objcet realm/zonegroup/zone if they are referenced by objectstores?

CephClients can also reference pools and could be dependents of Block/File/Object resources that create pools just as NFS can.

I think that we should probably implement dependent safety checks for all Rook resources if we are going to do them for some so that users aren't confused if they aren't protected in some cases when they are in others and so we don't have to elaborately document which cases are and which cases are not protected.

Comment (travisn) [reply design/ceph/block-ceph-delete-if-pool-data.md:9]: > Would you also include not deleting RBD mirroring CRs if block pools reference them?
> Or deleting objcet realm/zonegroup/zone if they are referenced by objectstores?

I was just providing examples, agreed there are other safety checks to add as you mentioned.

> CephClients can also reference pools and could be dependents of Block/File/Object resources that create pools just as NFS can.

While CephClients may be used for access to those resources, I don't see it as critical to detect. It's a weaker association that doesn't affect the ability to preserve data. Or at least there is a workaround to use the admin key if needed.

> I think that we should probably implement dependent safety checks for all Rook resources if we are going to do them for some so that users aren't confused if they aren't protected in some cases when they are in others and so we don't have to elaborately document which cases are and which cases are not protected.

Certainly, let's just agree on the goal. If it's potential data loss, we should protect against it. If there is some weaker association or there is a workaround, it may not be necessary.

File: design/resource-dependencies.md [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+- ObjectBucket (COSI)
+
+Kubernetes resources which can depend on Rook:
+- StorageClass
+- PersistentVolume
+
+## Proposed dependency relationships
+
+#### `CephCluster`
+Primary dependents which can create/delete pools:
```
Comment (travisn) [inline design/resource-dependencies.md:28]: All Rook CRs in the same namespace would be dependents of the CephCluster CR, right? Any of them should block the CephCluster removal. (edit: I see a comment about this at the end of the doc, if it could just be clarified here)

File: design/resource-dependencies.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  - `spec.storageClassName == <name of StorageClass which references the CephBlockPool>`
+
+Dependents via FlexVolume:
+- StorageClasses that have:
+- PersistentVolumes that have:
+
+TODO: decide if we should treat PVs OR StorageClasses as dependents. IMO, in following with the rest
+of the doc, a StorageClass that exists and references a pool is a good dependent and has a more
+defined structure than PV's `VolumeAttributed` arbitrary k-v list which can change due to changes in
+ceph-csi. Users will not be able to delete the StorageClass until all PVs which reference it are
```
Comment (travisn) [inline design/resource-dependencies.md:61]: I'm pretty sure you can delete a storage class even if there are PVs from it. I expect the storage class is only used at PV creation time.

File: design/resource-dependencies.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  - `spec.CSI.VolumeAttributes["clusterID"] == <ceph-cluster-namespace>` AND
+  - `spec.CSI.VolumeAttributes["pool"]` OR `spec.CSI.VolumeAttributes["journalPool"]` references a `CephBlockPool` pool
+  - OR
+  - `spec.storageClassName == <name of StorageClass which references the CephBlockPool>`
+
+Dependents via FlexVolume:
+- StorageClasses that have:
+- PersistentVolumes that have:
+
+TODO: decide if we should treat PVs OR StorageClasses as dependents. IMO, in following with the rest
```
Comment (travisn) [inline design/resource-dependencies.md:58]: Traversing PVs and storage classes feels out of scope to me. There could be thousands of PVs that we don't even own. Instead of querying PVs and storage classes, this is where it seems correct to query Ceph if there are related resources.

Comment (travisn) [reply design/resource-dependencies.md:58]: If we really needed to guarantee correctness, sure, traversing all those could make sense. That level of completeness just shouldn't be necessary for cluster cleanup IMO.

File: design/resource-dependencies.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+of the doc, a StorageClass that exists and references a pool is a good dependent and has a more
+defined structure than PV's `VolumeAttributed` arbitrary k-v list which can change due to changes in
+ceph-csi. Users will not be able to delete the StorageClass until all PVs which reference it are
+removed, which allows us to delegate that bit of work to k8s.
+
+#### `CephFilesystem`
+Dependents which can consume pools:
+- CephNFS in the same namespace that have `spec.pool == <a provider pool>`
+- CephClient in the same namespace that has any `spec.caps` value with the string `pool=<a provider pool>`
```
Comment (travisn) [inline design/resource-dependencies.md:67]: I wonder how often CephClient CRs are created for a specific pool. They aren't very common in the first place. I don't disagree about the dependency, it's lowers the priority to implement, for example so we can break the implementation up into multiple PRs.

File: design/resource-dependencies.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+- ObjectBucket
+
+#### Mirroring/multi-site resources
+CephRBDMirror has dependents:
+- CephBlockPools in the same namespace that have `spec.mirroring.enabled == true`
+
+CephFilesystemMirror has dependents:
+- CephFilesystems in the same namespace that have `spec.mirroring.enabled == true`
+
+CephObjectRealm has dependents:
```
Comment (travisn) [inline design/resource-dependencies.md:91]: How about moving the object-related CRs up with the object store section?

File: design/resource-dependencies.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+CephObjectZoneGroup has dependents:
+- CephObjectZones in the same namespace that have `spec.zoneGroup == CephObjectZoneGroup.metadata.name`
+
+CephObjectZone has dependents:
+- CephObjectStores in the same namespace that have `spec.zone.name == CephObjectZone.metadata.name`
+
+### Analysis
+We can identify some common patterns in determining whether a resource is a dependent of a given
+"provider" resource:
+- Does a dependent reference the current Rook-Ceph operator (notably for CSI/Flex)?
```
Comment (travisn) [inline design/resource-dependencies.md:103]: I hope we don't need to touch anything related to the flex driver. :)

File: design/resource-dependencies.md (hunk 9 above)
Comment (BlaineEXE) [reply design/resource-dependencies.md:67]: I agree with that. I'm finding this doc useful in order to think about how we can reuse pieces of code between controllers and types of dependencies.

File: design/resource-dependencies.md (hunk 7 above)
Comment (BlaineEXE) [reply design/resource-dependencies.md:61]: Oh! You're right. I guess we need to check both. :|

File: design/resource-dependencies.md (hunk 8 above)
Comment (BlaineEXE) [reply design/resource-dependencies.md:58]: Currently, we do list PVs when deleting the CephCluster but not StorageClasses.

Based on what we've been talking about and the scheme proposed here, I think I agree with you about PV traversal not being needed for cluster cleanup. Relatively speaking, I think it is _more_ important to check PVs when deleting CephBlockPools and CephFilesystems than when deleting CephClusters since those actually provision the type of storage which the PV will get and since the CephCluster has block/file as dependents. However, it is a good point that it could be a lot of work to traverse StorageClasses and PVs in large clusters.

I'm not sure querying Ceph would be able to tell us if there are PVs. We could check to see if there are images in a block pool or modifications to a filesystem, but that won't tell us if there are existing claims on block/file storage that haven't yet written data.

I don't think we **have** to search StorageClasses or PersistentVolumes to see if any reference storage when we are doing deletes. And we can definitely do this in a separate PR. I think it's worth discussing whether there is user benefit for Rook doing that traversal on deletion and whether it is appropriate for Rook to query those resources. 

For this doc, I would like to capture what is needed in order to inspect PV/SC dependencies so that we can see the complexity and weigh whether we want to include these as dependents in our design plans.

File: design/resource-dependencies.md (hunk 6 above)
Comment (BlaineEXE) [reply design/resource-dependencies.md:28]: Right! And that seems a good clarification.

File: cluster/examples/kubernetes/ceph/object-multisite.yaml [hunk 12]
```diff
@@ -2,44 +2,44 @@
...
-      requireSafeReplicaSize: true
-  dataPool:
-    failureDomain: host
-    replicated:
-      size: 1
-      requireSafeReplicaSize: true
-    parameters:
-      compression_mode: none
----
+# apiVersion: ceph.rook.io/v1
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/object-multisite.yaml:5]: You didn't intend to commit this file, right? looks like some local testing

File: design/ceph/block-ceph-delete-if-pool-data.md [hunk 13]
```diff
@@ -0,0 +1,146 @@
+# Blocking CephCluster deletion if user data is present
+
+
```
Comment (travisn) [inline design/ceph/block-ceph-delete-if-pool-data.md:3]: Some of the design docs have been adding a comment like this...
```suggestion
**Targeted for v1.7**
```

File: cluster/examples/kubernetes/ceph/object-multisite.yaml (hunk 12 above)
Comment (BlaineEXE) [reply cluster/examples/kubernetes/ceph/object-multisite.yaml:5]: Right....

**Comment by raghavendra-talur** [comment]
Thank you for starting this work @BlaineEXE ; when we worked on the uninstall story in the ocs-operator, we arrived at similar concerns and checks for the Ceph Resources. Most of the implementation was done in the ocs-operator and the knowledge of the dependencies were hard coded as a list [for example](https://github.com/openshift/ocs-operator/blob/b80f32df7b1984695066c6634e6e01f963d8d02d/controllers/storagecluster/uninstall_reconciler.go#L290). 

Bringing the dependency knowledge inside rook will help with two aspects:
1. It will be useful for everyone who uses rook, even those who are not using through ocs-operator
2. It will make the implementation more strict and enforce changes when the underlying architecture changes.

I have one question regarding the configuration `allowUninstallWithVolumes` setting on the CephCluster. This setting determines whether to uninstall the cluster when volumes still exist.  When it is set to `true`; rook ignores existing PVCs and does a forced uninstall. It would be good to have the interaction with setting also described in the design doc. You touch upon some aspects but not explicitly.

**Comment by raghavendra-talur** [comment]
cc: @iamniting

File: design/ceph/resource-dependencies.md [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+### Implementation in stages
+This design will result in changes to every major Rook-Ceph controller. However, it should be quite
+easy to tackle these changes in stages so that changes can be more easily implemented and reviewed.
+
+Stages by priority:
+1. Block deletion of a CephCluster when there are other Rook-Ceph resources in the same namespace
+1. Block deletion of a CephBlockPool, CephFilesystem, or CephObjectStore when a CephNFS uses one of its pools.
+1. Block deletion of a CephObjectStore when there is a user bucket present in the store.
+   - This already has an implementation, but it blocks for unrelated OBs
+1. Block deletion of a CephBlockPool or CephFilesystem when there is a PersistentVolume reliant on it.
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:255]: > What is the approach to detect filesystem PVs? 

This is defined under the CephFilesystem section above in the doc.

> What is the approach to detect filesystem PVs? Traversing all PVs in the cluster isn't scalable.

I expect we can use field selectors to list a subset of PVs which match some of the known parameters we need to check. We can do further checking on that list if needed.

File: design/ceph/resource-dependencies.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+`CephObjectZoneGroup` has dependents:
+- CephObjectZones in the same namespace that have `spec.zoneGroup == CephObjectZoneGroup.metadata.name`
+
+#### Mirroring resources
+CephRBDMirror has dependents:
+- CephBlockPools in the same namespace that have `spec.mirroring.enabled == true`
+
+CephFilesystemMirror has dependents:
+- CephFilesystems in the same namespace that have `spec.mirroring.enabled == true`
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:156]: I'm not sure what you mean, though you would know best since you are implementing CephFilesystemMirror. From what I can tell, CephFilesystemMirror can't mirror block pools, so I don't think this subsequently follows. It would be good to talk about this specific detail in tomorrow's huddle.

File: design/ceph/resource-dependencies.md [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+- CephFilesystems in the same namespace
+- CephObjectStores in the same namespace
+- CephObjectZones in the same namespace
+
+Dependents which can consume arbitrary pools including user-created pools:
+- CephNFSes in the same namespace
+- CephClients in the same namespace
+
+Dependents that do not interact with pools:
+- CephRBDMirrors in the same namespace
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:71]: Yes! I had to update my diagram also.

File: design/ceph/resource-dependencies.md [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+#### `CephCluster`
+A CephCluster does not create pools itself, but the Ceph cluster it represents houses pools, and
+users can manually create pools using Ceph tooling outside of Kubernetes manifests. It is useful but
+not critical to understand which resources interact with pools and in what ways.
+
+Dependents which can create/delete pools:
+- CephBlockPools in the same namespace
+- CephFilesystems in the same namespace
+- CephObjectStores in the same namespace
+- CephObjectZones in the same namespace
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:64]: CephObjectZone CRD has a pool spec.

File: design/ceph/resource-dependencies.md [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+The main downside to this approach is that Rook will not be able to report if there are specific
+OBs or COSI Buckets consuming storage from the CephObjectStore. An admin would need to examine the
+resources in their cluster to determine if there are claims to the storage manually. A diligent
+admin will likely have done this work beforehand.
+
+This document proposes to implement this strategy to avoid reliance on external APIs. Rook
+developers should revisit the decision at a later date to discuss whether the strategy is continuing
+to adequately meet users' needs and whether the drawbacks noted are causing any issues.
+
+### Implementation in stages
```
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:246]: I think the summary, in my mind, is the proposed dependency relationships and the diagram. This is merely a detail about when things will be implemented.

File: design/ceph/resource-dependencies.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (leseb) [reply design/ceph/resource-dependencies.md:158]: Right, CephFilesystemMirror is not mirroring "pool" but it depends on CephFilesystem which depends on CephBlockPool (for fs data and fs metadata).

File: design/ceph/resource-dependencies.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/resource-dependencies.md:248]: There are a lot of detailed dependencies and we aren't planning to enforce them all, right? For me at least, it's helpful to have a "tldr" section at the top on what the proposal is before diving into the details. But as long as we have this section somewhere in the doc, I'll be content. :)

File: design/ceph/resource-dependencies.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/resource-dependencies.md:257]: > I expect we can use field selectors to list a subset of PVs which match some of the known parameters we need to check. We can do further checking on that list if needed.

Listing PVs still doesn't seem scalable. Instead, can we query ceph for subvolumes in the filesystem?

[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/resource-dependencies.md:257]: >> We already check for this when deleting the pool if there are any rbd images.

> We check this but only when the CephCluster is deleted and not when CephBlockPools are deleted. We should do this when deleting CephBlockPools and let the CephCluster implicitly check for this by having it check for CephBlockPools.

Where are the rbd images checked when the cluster is deleted? I only see this in the pool controller, which calls `DeletePool()`, which calls `checkForImagesInPool()`, which fails if there are any images.

File: design/ceph/resource-dependencies.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:158]: I don't think that's right. The CephFilesystem doesn't require (nor allow) using an existing pool. Therefore, it is cannot be dependent on CephBlockPool. The CephFilesystem controller creates its own pools when it needs them. It still uses the Pool spec, but it is copied and not directed toward a CephBlockPool.

File: design/ceph/resource-dependencies.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:257]: > > > We already check for this when deleting the pool if there are any rbd images.
> 
> > We check this but only when the CephCluster is deleted and not when CephBlockPools are deleted. We should do this when deleting CephBlockPools and let the CephCluster implicitly check for this by having it check for CephBlockPools.
> 
> Where are the rbd images checked when the cluster is deleted? I only see this in the pool controller, which calls `DeletePool()`, which calls `checkForImagesInPool()`, which fails if there are any images.

See the `checkIfVolumesExist` code added here: https://github.com/rook/rook/pull/6145/files

File: design/ceph/resource-dependencies.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:248]: In my estimation, we should either enforce them all, or only enforce dependencies from the `CephCluster`. Otherwise, users may be confused when they are protected from some dependencies but not others, and it will be difficult to document the minutiae of what is and what is not protected. 

If it is all protected, then we can just document that Rook resources should be protected from being deleted if there are other Rook resources referencing them, and if they aren't it can be considered a bug. Additional protections include buckets created by users, lib-bucket-provisioner OBCs, or COSI Buckets as well as PersistentVolumes that reference Rook storage.

[the remark this answers is not in the document]
Comment (BlaineEXE) [reply design/ceph/resource-dependencies.md:265]: I created a new summary section at the top.

File: design/ceph/resource-dependencies.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (leseb) [reply design/ceph/resource-dependencies.md:175]: Ok, I see you are treating that case differently because `CephFilesystem` only uses the spec and not the CRD.

**Review by leseb — APPROVED** [review]

**Comment by BlaineEXE** [comment]
Adding do-not-merge so I can add sections discussed in Rook community meeting today:
- include status/events to notify users when resources won't be deleted
- add future work to include dependency checking in admission controller

File: design/ceph/resource-dependencies.md [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+blocking the deletion. This design proposes that the Rook operator report to the user when it is
+blocking deletion of a resource due to dependents in two ways:
+- report it in a status condition of the resource object being deleted
+- report a Kubernetes Event ([reference](https://kubernetes.io/blog/2018/01/reporting-errors-using-kubernetes-events/))
+
+Using both of these methods will maximize user visibility.
+
+**Status:**
+Reported statuses will be modified as follows:
+1. Object's status.phase should be changed to "Deleting" as soon as a deletion timestamp is detected
```
Comment (travisn) [inline design/ceph/resource-dependencies.md:200]: a `kubectl get <CR>` should already show it deleted, but updating the phase certainly wouldn't hurt if we're already reporting conditions and events

File: design/ceph/resource-dependencies.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (travisn) [reply design/ceph/resource-dependencies.md:303]: Ok I guess it's already implemented :)

**Review by travisn — APPROVED** [review]
